optuna/integration/sklearn.py

```diff
@@ -0,0 +1,925 @@
+
+        try:
+            self.best_estimator_.set_params(**self.study_.best_params)
+        except ValueError as e:
+            logger.exception(e)
+
+        logger.info(
+            'Refitting the estimator using {} samples...'.format(n_samples)
+        )
+
+        start_time = time()
+
+        self.best_estimator_.fit(X, y, **fit_params)
+
+        self.refit_time_ = time() - start_time
+
+        logger.info(
+            'Finished refitting! '
+            '(elapsed time: {:.3f} sec.)'.format(self.refit_time_)
+        )
+
+        return self
+
+    def fit(
+        self,
+        X,  # type: TwoDimArrayLikeType
+        y=None,  # type: Optional[Union[OneDimArrayLikeType, TwoDimArrayLikeType]]
+        groups=None,  # type: Optional[OneDimArrayLikeType]
+        **fit_params  # type: Any
+    ):
+        # type: (...) -> 'OptunaSearchCV'
+        """Run fit with all sets of parameters.
+
+        Args:
+            X:
+                Training data.
+
+            y:
+                Target variable.
+
+            groups:
+                Group labels for the samples used while splitting the dataset
+                into train/test set.
+
+            **fit_params:
+                Parameters passed to ``fit`` on the estimator.
+
+        Returns:
+            self:
+                Return self.
+        """
+
+        self._check_params()
+
+        random_state = check_random_state(self.random_state)
+        max_samples = self.subsample
+        n_samples = _num_samples(X)
+        old_level = logger.getEffectiveLevel()
+
+        if self.verbose > 1:
+            logger.setLevel(DEBUG)
+        elif self.verbose > 0:
+            logger.setLevel(INFO)
+        else:
+            logger.setLevel(WARNING)
+
+        self.sample_indices_ = np.arange(n_samples)
+
+        if type(max_samples) is float:
+            max_samples = int(max_samples * n_samples)
+
+        if max_samples < n_samples:
+            self.sample_indices_ = random_state.choice(
+                self.sample_indices_,
+                max_samples,
+                replace=False
+            )
+
+            self.sample_indices_.sort()
+
+        X_res = safe_indexing(X, self.sample_indices_)
+        y_res = safe_indexing(y, self.sample_indices_)
+        groups_res = safe_indexing(groups, self.sample_indices_)
+        fit_params_res = fit_params
+
+        if fit_params_res is not None:
+            fit_params_res = {
+                key: _index_param_value(
+                    X,
+                    value,
+                    self.sample_indices_
+                ) for key, value in fit_params.items()
+            }
+
+        classifier = is_classifier(self.estimator)
+        cv = check_cv(self.cv, y_res, classifier)
+
+        self.n_splits_ = cv.get_n_splits(X_res, y_res, groups=groups_res)
+        self.scorer_ = check_scoring(self.estimator, scoring=self.scoring)
+
+        if self.study is None:
```

Comment:
How about moving study initialization to `__init__`, so that the user can call `set_user_attr` before fitting.

Reply:
Let's forget about this matter for this PR.

According to the sklearn contribution guide, any parameter validation should not be done in `__init__`.